runtime/stdlib/cadence_v0.42_to_v1_contract_upgrade_validator.go

```diff
@@ -338,11 +338,31 @@ typeSwitch:
 			return nil
 		}
 	case *ast.IntersectionType:
-		// intersection types cannot be upgraded unless they have a legacy restricted type,
-		// in which case they must be upgraded according to the migration rules: i.e. R{I} -> R
+		// If the intersection type have doesn't a legacy restricted type,
+		// the interface set must be compatible.
 		if oldType.LegacyRestrictedType == nil {
-			break
+			foundIntersectionType, ok := newType.(*ast.IntersectionType)
+			if !ok {
+				return newTypeMismatchError(oldType, newType)
+			}
+
+			if len(oldType.Types) != len(foundIntersectionType.Types) {
+				return newTypeMismatchError(oldType, newType)
+			}
+
+			for index, expectedIntersectedType := range oldType.Types {
+				foundType := foundIntersectionType.Types[index]
```

Comment:
Does that mean the types have to occur in exactly the same order as before?
Definitely OK, but we might need to help developers here in the case where they are not, because e.g. for a case like `{A, B}` -> `{B, A}`, this would fail (right?), but it might not be clear the reason is the order.

Reply:
ah good point! I copied the logic form the existing type-comparator. But yeah, order shouldn't matter now.

Reply:
added with https://github.com/onflow/cadence/pull/3283/commits/2d443446cae94474b68472441b9e43e557171d00